Convert Notes to use option_group instead of hardcoded list

colemanw opened the pull request:
Overview
----------------------------------------
Allows Notes to be added to different types of records by extensions.

Before
----------------------------------------
Hard-coded list of entities allowed to contain Notes.

After
----------------------------------------
Same list, but now it can be easily modified by extensions.

colemanw commented:
Marking WIP until regen.sh is run. I have a hard time getting that to work locally.

monishdeb commented:
@colemanw tested the patch in my local, works fine - correctly replaces the hardcoded list with option group `note_used_for`

Shouldn't the patch also include the DAO changes? 
```diff
diff --git a/CRM/Core/DAO/Note.php b/CRM/Core/DAO/Note.php
index 35067ea292..2b6da8592c 100644
--- a/CRM/Core/DAO/Note.php
+++ b/CRM/Core/DAO/Note.php
@@ -180,7 +180,8 @@ class CRM_Core_DAO_Note extends CRM_Core_DAO {
           'bao' => 'CRM_Core_BAO_Note',
           'localizable' => 0,
           'pseudoconstant' => [
-            'callback' => 'CRM_Core_BAO_Note::entityTables',
+            'optionGroupName' => 'note_used_for',
+            'optionEditPath' => 'civicrm/admin/options/note_used_for',
           ],
           'add' => '1.1',
         ],
  ```
  
  And should I also target other table's entity_table to replace the hard-coded list with optiongroup esp those fields which got dynamic list like civicrm_entity_batch.entity_table , civicrm_financial_trxn.entity_table etc. (but won't consider LineItem.entity_table that should be restricted to few table names)

colemanw commented:
Good catch @monishdeb. I've rebased this PR with the addition of the DAO file updates.

monishdeb commented:
Cool. Yep am happy with the final patch.

colemanw commented:
I'm not sure about converting other option groups. @totten said this about it on [the dev channel](https://chat.civicrm.org/civicrm/pl/gdwr88ekdpg8jcrobjusf3nj9w):

> The list of note-addable things could be:
>
> 1. An option-group
> 2. A hookable list
> 3. A fixed list
>
> Each one has a sort of "sweet spot"/"ideal use case". IMHO:
>
> 1. OG/OV-list - Works nicely if it should be curated by power-users/admins
> 2. Hook-list - Works nicely if it should be curated by developers
> 3. Fixed-list - Works nicely if changes can have systemic/ripple effects across components (*though in practice that rationale doesn't come up much... fixed-lists are usually used because they're easier/less administrativa*)
>
> (*That's just a reflection of status-quo. If one made improvements... say,  better ordering in `hook_managed`, easier mechanism for declaring OGs+OVs, or more ways to mix styles... then the calculus would change.*)

monishdeb commented:
Oh ok, thanks for sharing the excerpt.

demeritcowboy commented on xml/templates/civicrm_data.tpl:
```diff
@@ -755,6 +757,12 @@ VALUES
    (@option_group_id_tuf, '{ts escape="sql"}Attachments{/ts}', 'civicrm_file',     'File',     NULL, 0, NULL, 4, NULL, 0, 0, 1, NULL, NULL, NULL),
 
+-- Note used for
+   (@option_group_id_nuf, 'Contact',      'civicrm_contact',      'Contact',      NULL, 0, NULL, 1, NULL, 0, 0, 1, NULL, NULL, NULL),
```
I think you need {ts} here like the group above.

demeritcowboy commented on CRM/Upgrade/Incremental/php/FiveForty.php:
```diff
@@ -51,9 +51,38 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
+    foreach ($values as $value => $name) {
+      \CRM_Core_BAO_OptionValue::ensureOptionValueExists([
+        'value' => $value,
+        'name' => $name,
```
Similarly you'd want to explicitly specify label here with a ts, but not ts($name) since that's not good - maybe could expand the array above to also have entries for ts('Contact'), ts('Relationship'),.. ?

colemanw commented:
@demeritcowboy thanks for the review. I'm not sure how much it matters in this case to have translated labels because they are never displayed in the UI. But for the sake of consistency, sure. I've updated both spots.

colemanw commented:
@demeritcowboy Uh oh, now the regen script doesn't want to run for some reason. It says it can't apply the patch, which is odd since Github says no conflicts. https://github.com/colemanw/civicrm-regen/runs/2851777080?check_suite_focus=true

colemanw commented:
Ok looks like that was due to the checksum for `CRM_Contribute_DAO_ContributionProduct` needing to be regenerated. I've left it out of this PR so as not to cause merge conflicts.

demeritcowboy commented:
Oh ok so you're saying they don't even show up at civicrm/admin/options?reset=1. The "Tag Used For" ones do but ok either way then.
